### ISSUE TITLE
<format>: Format string parsing needs to be charset aware

### ISSUE THREAD

**Issue by statementreply** [microsoft/STL#1576]
Per discussion on Discord.

Some multibyte charsets use 0x7b and 0x7d (the same as `'{'` and `'}'`, respectively) for the second byte of character encoding. We need to be aware of this issue when implementing #30, and properly parse format strings containing such characters.

For example, the string `"日本地図"` (meaning "map of Japan" in Japanese) is encoded as `"\x93\xfa\x96\x7b\x92\x6e\x90\x7d"` in Shift JIS (code page 932), which contains both 0x7b and 0x7d. When running under code page 932, it shouldn't be parsed as `"\x93\xfa\x96"` + **`"{\x92n\x90}"`**.

AFAIK, code pages 932, 936, 950, and 54936 contain such encodings.

**Command-line test case**

```
D:\Temp>type format_sjis.cpp
```

```cpp
#include <format>
#include <iostream>
#include <string_view>

using namespace std;

int main() {
    constexpr auto str = "\x93\xfa\x96\x7b\x92\x6e\x90\x7d"sv;
    cout << str << "\n";
    cout << format(str) << "\n";
}
```

```
D:\Temp>chcp 932
Active code page: 932

D:\Temp>cl /EHsc /W4 /WX /std:c++latest format_sjis.cpp
[...]

D:\Temp>.\format_sjis.exe
[...]
```

**Expected behavior**

Should print:
```
日本地図
日本地図
```

**Comment by statementreply** [comment]
Additionally, what should we do with compile time checking if WG21-P2216 were voted in?

**Comment by CaseyCarter** [comment]
Thank you for the concrete examples! I had concerns about this earlier, but didn't know of any actual encodings that used `{`/`}`.

> Additionally, what should we do with compile time checking if [WG21-P2216](https://wg21.link/P2216) were voted in?

Implement it, which I suspect will be equivalent to writing a compiletime `mbtowc` =(

**Comment by barcharcraz** [comment]
I've run into this issue already (I think you pointed it out @statementreply), and as far as we can tell as long as we don't narrow wchar_t format strings things will work, if you don't mix TUs of different source charsets.

**Comment by barcharcraz** [comment]
Fir P2216 I'm not super worried about the compile time format string checking, all the parsing is constexpr anyways, so it's just a minor API change. The binary size "fix" is a little more problematic.

**Comment by CaseyCarter** [comment]
This should have been closed by #1815.
